src/pgwire/src/protocol.rs

```diff
@@ -83,6 +87,11 @@ pub struct RunParams<'a, A> {
     /// Whether this is an internal server that permits access to restricted
     /// system resources.
     pub internal: bool,
+    /// During handling the query, did we increment the connection count. Used
+    /// during connection teardown to determine if we need to decrement the count.
+    pub incremented_connection_count: &'a mut bool,
```

Comment:
I didn't try to fully understand what was going on here (ran out of time tonight), but this seems a bit scary and maybe there's a better design. I'll take a look tomorrow try to make a suggestion.

Reply:
Yeah, lemme know. Another idea i had was to protocol.rs's `run` return a `Result<bool, io::Error>` where the bool indicates whether or not we need to decrement the connection count. I was thinking of keeping a bool in that function and changing all the returns to `return blah.await.map(|()| should_decrement_count)` but that seemed wordy

Reply:
Ohh oh, this is per connection. Due to the `internal` flag I thought this was the singleton for the whole server and it was sharing this state back and forth like during connection startup or something. I'm ok with this.